From 27c33f7d66eba4ff9574bb4b4d945c431a2e04b3 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:05:13 -0400
Subject: [usb] cypress_m8: only wake up MSR queue on changes

Message-id: <1410887463-20674-242-git-send-email-dzickus@redhat.com>
Patchwork-id: 95314
O-Subject: [RHEL7 PATCH 241/591] USB: cypress_m8: only wake up MSR queue on changes
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 7603381e3e7ffebe8c0b46c416fe0f88e13b4d34
Author: Johan Hovold <jhovold@gmail.com>
Date:   Thu Jan 2 22:49:26 2014 +0100

    USB: cypress_m8: only wake up MSR queue on changes

    Only wake up MSR wait queue on actual modem-status changes.

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/cypress_m8.c b/drivers/usb/serial/cypress_m8.c
index 558605d..07e0033 100644
--- a/drivers/usb/serial/cypress_m8.c
+++ b/drivers/usb/serial/cypress_m8.c
@@ -1187,7 +1187,10 @@ static void cypress_read_int_callback(struct urb *urb)
  if (priv->current_status != priv->prev_status) {
   priv->diff_status |= priv->current_status ^
    priv->prev_status;
-  wake_up_interruptible(&port->port.delta_msr_wait);
+
+  if (priv->diff_status & UART_MSR_MASK)
+   wake_up_interruptible(&port->port.delta_msr_wait);
+
   priv->prev_status = priv->current_status;
  }
  spin_unlock_irqrestore(&priv->lock, flags);
diff --git a/drivers/usb/serial/cypress_m8.h b/drivers/usb/serial/cypress_m8.h
index 68bfd51..119d2e1 100644
--- a/drivers/usb/serial/cypress_m8.h
+++ b/drivers/usb/serial/cypress_m8.h
@@ -66,6 +66,7 @@
 #define CONTROL_RTS 0x10 /* request to send */
 #define CONTROL_RESET 0x08 /* sent with output report */
 
+#define UART_MSR_MASK 0xf0
 #define UART_RI  0x80 /* ring indicator */
 #define UART_CD  0x40 /* carrier detect */
 #define UART_DSR 0x20 /* data set ready */
-- 
1.7.1